qlib/CsvUtil.qm

```diff
@@ -117,8 +117,80 @@ while (i.next())
 
     Use the \c "header_lines" and \c "header_names" @ref abstractcsviterator_options "options" to automatically read the header names from the file if present.  Use the \c "fields" @ref abstractcsviterator_options "option" to describe the fields and perform transformations on the data read.  For more information, see the @ref CsvUtil::CsvFileIterator "CsvFileIterator" class.
 
+    @anchor multi_type
+    Multi-type lines are supported as well. Multi-type means the line may have different structure, i.e.fields. The type of line is evaluated in run-time using \c "resolve_type" @ref abstractcsviterator_options "option", typically according to the value of the first field at line. It works for both @ref CsvUtil::AbstractCsvIterator "AbstractCsvIterator" and @ref CsvUtil::AbstractCsvWriter "AbstractCsvWriter".
+    @code{.py}
+#!/usr/bin/env qore
+
+%new-style
+
+%requires CsvUtil
+
+const opts = (
+    "fields": (
+        "type": "int",
+
+        "invoice_no": "string",
+        "date": ("type": "date", "format": "YYYYMMDD"),
+
+        "item_no": "string",
+        "item": "string",
+        "pcs": "int",
+        "price": "number",
+
+        "total": "number",
+    ),
+    "headers": (
+        "header": ("type", "invoice_no", "date"),
+        "item":   ("type", "item_no", "item", "pcs", "price"),
+        "footer": ("type", "total"),
+    ),
+    'resolve_type': softstring sub(hash values) {
+        softint type = values{'type'}??values{'0'};
+        switch (type) {
+            case 1:
+                return "header";
+            case 2:
+                return "item";
+            case 3:
+                return "footer";
+        }
+        return NULL;
+    },
+);
+
+CsvFileIterator i("example-multi-file.csv", opts);
+CsvFileWriter writer("example-multi-file-copy.csv", ("write-headers": False, "optimal-quotes" : True, "quote_escape" : '"')+opts);
+while (i.next()) {
+    printf("%d: %y\n", i.index(), i.getValue());
+    writer.writeLine(i.getValue());
+}
+    @endcode
+
+
+    If \c "example-multi-file.csv" is:
+    @verbatim
+1,2016-01,20160401
+2,123,"Sony, Xperia S",1,100
+2,124,"Nokia, Lumia",2,150
+3,250
+    @endverbatim
+
+    The data is read verbatim, each value is returned as a string, header names are generated numerically; the output is:
+    @verbatim
+1: ("type": 1, "invoice_no": "2016-01", "date": 2016-04-01)
+2: ("type": 2, "item_no": 123, "item": "Sony, Xperia S", "pcs": 1, "price": 100)
+3: ("type": 2, "item_no": 124, "item": "Nokia, Lumia", "pcs": 2, "price": 150)
+4: ("type": 3, "total": 250)
+    @endverbatim
+
+    Also the \c "example-multi-file-copy.csv" will contain data from the original file formatted as CSV.
+
     @section csvutil_relnotes Release Notes
 
+    @subsection csvutil_v1_6 Version 1.6
```

Comment:
please put these release notes in v 1.5 - v 1.5 is the current unreleased version where all changes have gone so far - we only increment the release number once for each a public release